### PR TITLE
8252003: remove usage of PropertyResolvingWrapper in vmTestbase/nsk/jvmti

### PR DESCRIPTION
**Pull request by iignatev**
Hi all,

could you please review the patch which removes `PropertyResolvingWrapper` from `vmTestbase/nsk/jvmti` tests? as `jtreg` doesn't support spaces in the arguments and doesn't handle `"` in any special ways, the patch also:
 - `s/"-javaOpts=/-javaOpts="/`
 - makes `nsk.jvmti.scenarios.general_functions.GF08` to use 2nd arg as `verboseType` and 3rd and the rest args concatenated as `phrase` and updates the tests accordingly
 - removes spaces and surrounding `"` from `nsk.jvmti.test.property*`
 - removes `"` surrounding `-agentlib:`, replaces spaces in `-agentlib` with `,` and updates `ArgumentHandler` to treat `,` (as well as ` ` and `~`) as options delimiters, so it's consistent w/ `jvmti_tools.cpp`

testing: ✅  `vmTestbase/nsk/jvmti` on {linux,windows,macos}-x64

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252003](https://bugs.openjdk.java.net/browse/JDK-8252003): remove usage of PropertyResolvingWrapper in vmTestbase/nsk/jvmti

### Reviewers
 * [Chris Plummer](https://openjdk.java.net/census#cjplummer) (@plummercj - **Reviewer**) ⚠️ Review applies to 6190e55b1d51e024bce14622ba510c75dc35fcb6
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/370/head:pull/370`
`$ git checkout pull/370`

### CODE REVIEW

**Comment by iignatev** [comment]
ping?

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/general_functions/GF08/gf08t003/TestDescription.java [hunk 1]
```diff
@@ -58,7 +58,7 @@
  *      nsk.jvmti.scenarios.general_functions.GF08.gf08t
  *      gf08t003
  *      nsk.jvmti.scenarios.general_functions.GF08.gf08t003
- *      "Registering JNI native method"
  *      jni
+ *      Registering JNI native method
```
Comment (plummercj) [inline test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/general_functions/GF08/gf08t003/TestDescription.java:61]: The motivation for this change isn't obvious to me, and seems less readable than the original code.

File: test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/ArgumentHandler.java [hunk 2]
```diff
@@ -154,7 +154,7 @@ protected void parseOptionString(String optionString) {
         if (optionString == null)
             return;
 
-        StringTokenizer st = new StringTokenizer(optionString);
+        StringTokenizer st = new StringTokenizer(optionString, " ~,");
```
Comment (plummercj) [inline test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/ArgumentHandler.java:157]: I don't see jvmti_tools.cpp changes as part of this PR. Can you elaborate on why this change is needed now.

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/general_functions/GF08/gf08t003/TestDescription.java (hunk 1 above)
Comment (iignatev) [reply test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/general_functions/GF08/gf08t003/TestDescription.java:61]: the reason I had to this change is that jtreg doesn't handle `"` in any special way, so `gf08t003 nsk.jvmti.scenarios.general_functions.GF08.gf08t003 "Registering JNI native method" jni` is passed to `gf08t` as
[`gf08t003`, `nsk.jvmti.scenarios.general_functions.GF08.gf08t003`, `"Registering`, `JNI`, `native`, `method"`, `jni`]. the way I fixed that was by changing the order of `phrase`, `verboseType`, so `verboseType` is now 2nd arg, and `phrase` 3rd and the rest of args joined.

the alternative way would be to teach `gf08t` about `"` and make it join arguments until it gets an even number of `"` and remove them, this, from my PoV, would be a bit more confusing. if you however are of string opinion here, I can go w/ that alternative (or any other if you have more suggestions)

File: test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/ArgumentHandler.java (hunk 2 above)
Comment (iignatev) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/ArgumentHandler.java:157]: `jvmti_tools.cpp` wasn't changed by this PR, `jvmti_tools.cpp` always accepted ` `, `~` and/or `,` as delimiters[[*]]. as jtreg splits arguments by space symbol, we can't use ` ` in `-agentlib:`, so we had to change the delimiter used there to either `,` or `~`, `nsk/share/jvmti/ArgumentHandler` takes these options (via `getAgentOptionsString`) and then parse in `parseOptionString`. but `parseOptionString` was using `StringTokenizer` w/ default delimiters, ie ` \t\n\r\f`, and as a result it wasn't splitting options correctly after we changed the delimiter used in tests' `-agentlib`. in other words, this change is just fixing an old bug which we didn't care before as we always used ` ` in this agent's options. 

[*]: https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/jvmti_tools.cpp#L224

Comment (plummercj) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/ArgumentHandler.java:157]: I guess a better question would be is where is ~ being used in the arguments?

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/general_functions/GF08/gf08t003/TestDescription.java (hunk 1 above)
Comment (plummercj) [reply test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/general_functions/GF08/gf08t003/TestDescription.java:61]: So what code used to handle passing quoted arguments? Is this something that PropertyResolvingWrapper used to do?

Comment (iignatev) [reply test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/general_functions/GF08/gf08t003/TestDescription.java:61]: right, `PropertyResolvingWrapper` was joining arguments which are within `"`

File: test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/ArgumentHandler.java (hunk 2 above)
Comment (iignatev) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/ArgumentHandler.java:157]: not in the tests which use `nsk/share/jvmti/ArgumentHandler`, otherwise, we would see these tests failing as wouldn't parse them correctly. quick grepping showed that that 
 * `~` is used in `vm/mlvm/*/jvmti` tests, but these tests don't use `nsk/share/jvmti/ArgumentHandler`, they use `nsk/share/ArgumentParser/IgnoreUnknownArgumentParser` which doesn't parse agent options string at all
 * `,` has been used in `nsk/jvmti/scenarios/contention/TC05/tc05t001/` before this patch, and the test wasn't getting options right, it was setting `verbose,waittime` equal to `5`, instead of `verbose` to an empty string and `waittime` to `5`.

Comment (plummercj) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/ArgumentHandler.java:157]: Ok.

**Review by plummercj — APPROVED** [review]

File: test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/ArgumentHandler.java (hunk 2 above)
Comment (iignatev) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/ArgumentHandler.java:157]: after writing all of this, I think it makes sense to separate the fix in `ArgumentHandler` for the sake of clarity and easy of backporting. @plummercj, what do you think?

Comment (iignatev) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/ArgumentHandler.java:157]: separated into #438 / [8253872](https://bugs.openjdk.java.net/browse/JDK-8253872)

Comment (plummercj) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/ArgumentHandler.java:157]: I'd prefer that.

**Comment by sspitsyn** [comment]
@iignatev ,
What are the meaning of these properties ?:
  -Dnsk.jvmti.test.property=value_of_nsk.jvmti.test.property
  -Dnsk.jvmti.test.property=initial_value_of_nsk.jvmti.test.property
  -Dnsk.jvmti.test.property.empty.new=initial_value_of_nsk.jvmti.test.property.empty.new

Could you, please, explain this change? :
... scenarios/general_functions/GF08/gf08t001/TestDriver.java

@@ -87,8 +87,8 @@ public static void main(String[] args) throws Exception {
        nsk.jvmti.scenarios.general_functions.GF08.gf08t.main(new String[] {
                "gf08t001",
                nsk.jvmti.scenarios.general_functions.GF08.gf08t001.class.getName(),
 -               keyPhrase,
 -              "gc"});
 +              "gc",
 +              keyPhrase});
    }
}

**Review by sspitsyn — APPROVED** [review]

**Comment by iignatev** [comment]
> What are the meaning of these properties ?:
> -Dnsk.jvmti.test.property=value_of_nsk.jvmti.test.property
> -Dnsk.jvmti.test.property=initial_value_of_nsk.jvmti.test.property
> -Dnsk.jvmti.test.property.empty.new=initial_value_of_nsk.jvmti.test.property.empty.new

these are properties used by jvmti tests, the tests compare the actual value w/ expected values hardcoded in the test code. the purpose of these properties is different in different tests, ranging from just checking values by an agent to changing the value by the agent and checking that the value was changed.
 
> 
> Could you, please, explain this change? :
> 
> 
``` ... scenarios/general_functions/GF08/gf08t001/TestDriver.java
@@ -87,8 +87,8 @@ public static void main(String[] args) throws Exception {
        nsk.jvmti.scenarios.general_functions.GF08.gf08t.main(new String[] {
                "gf08t001",
                nsk.jvmti.scenarios.general_functions.GF08.gf08t001.class.getName(),
 -               keyPhrase,
 -              "gc"});
 +              "gc",
 +              keyPhrase});
    }
}
```

it's the same changes I described [above](https://github.com/openjdk/jdk/pull/370#discussion_r497655837), in two words, due to how jtreg handles spaces and `"` in the arguments, the positions of `phrase` and  `verboseType` have been switched.

**Comment by sspitsyn** [comment]
Okay, thanks!

**Comment by iignatev** [comment]
Serguei, Chris, thanks a lot for your reviews.

/integrate
